data/blog/how-to-refactor-and-not-break-things.md

```diff
@@ -0,0 +1,101 @@
+
+## Why We Wanted to Do This
+
+When doing a big refactoring like this the WHY is very important. We do not want to refactor just for refactoring's sake. 
+In recent years [OpenTelementry](https://opentelemetry.io/) has become more popular and we wanted to make sure Sentry is compatible with OpenTelemetry. 
+We discovered that with our current Hub implementation that was not the case, so we decided to remove the Hub and switch to have only Scopes and make them behave like Contexts in OpenTelementry. This makes Sentry 100% compatible with OpenTelemetry and paves the way for the future.
+
+## This Is a Huge Undertaking, We need a Plan!
+
+Some goals that we wanted to achieve with the refactoring:
+
+- Be as backward compatible as possible. If possible, our users should not need to change their custom Sentry code.
+- Because we changed how we handle data, we wanted to make sure that we do not use more resources (CPU and memory) compared to before the refactoring.
+- Do not break things. Behavior must stay the same.
+- Do it in baby steps. Have only PRs of manageable size. (No one can review a PR with 120 files changed.)
+- Because this is a massive change, we decided to make it a major version update. Sentry SDK 2.0!
+
+## Phase I: Preparation
+
+There is a quote by [Kent Beck](https://twitter.com/kentbeck/status/250733358307500032):
+
+> Make the change easy, then make the easy change. (Warning, the first part might be hard)
+
+In this first phase, we moved existing functionality away from the Hub into the Scope. We did several PRs that each moved one or two functions from the Hub into the Scope and changed the function in the Hub to call its counterpart in the Scope.
+
+This way we had a couple of smaller PRs that each dealt with one topic and were easy to review.
+
+After we moved all the functionality from the Hub into the Scope our extensive test suite was still all green, giving us confidence that we did not change any behavior. Because neither the top-level Sentry API nor the Hub-based API was changed this could be released in a minor version update.
+
+This concluded the preparation of our canvas. Time to make the change.
+
+## Phase II: New Scopes
+
+Phase I left us with a hollow shell of a Hub and all functionality in the Scope. We now refactored the Scope. This was the biggest part. We changed how the Scope was stored in memory (it's not on the Hub anymore but saved as a Python Context Variable). We also introduced three different flavors of the Scope for better encapsulation of data. If you want to dig into details read our [develop docs on the topic](https://develop.sentry.dev/sdk/hub_and_scope_refactoring/).
+
+After we updated the Scope the old Hub-based API could still be used, but under the hood, the new Scopes-based API was called. We did not release this phase because we wanted to first do Phase III.
```

Comment:
> We did not release this phase because we wanted to first do Phase III.

This doesn't really explain why not. It wouldn't add value? Would it at least help make sure there were no bugs there? (beyond what the test suite can do of course)

Reply:
true. good point. will rephrase.